JavaCollections2_task1/src/main/java/org/Marta/MyMap.java

```diff
@@ -0,0 +1,115 @@
+package org.Marta;
+
+public class MyMap implements OwnMap {
+    Node root;
+
+    public Node putRecursively(Node current, String key, String value) {
+        if (current == null) {
+            return new Node(key, value);
+        }
+        if (key.compareTo(current.getKey()) < 0) {
+            current.left = putRecursively(current.left, key, value);
+        } else if (key.compareTo(current.getKey()) > 0) {
+            current.right = putRecursively(current.right, key, value);
+        } else {
+            return current;
+        }
+        return current;
```

Comment:
główną zaletą mapy jest to, że dodawanie jest w czasie stałym czyli np. w bazie jest 1000000 rekordów czy 100000000000000000000000 - bez różnicy, tak samo szybko dodamy element.

W twoim przypadkuim istnieje szansa, że bedziemy musieli przechodzić **po każdym elemencie.** Czyli w przypadku gdy w rosnie rozmiar mapy - operacja dodawania trwa dłużej.

Reply:
a co z wyszukiwaniem? ;) w przypadku tablicy wyszukiwanie może wymagać przejścia nawet całej tablicy, a w przypadku zrównoważonego drzewa jest to złożoność logarytmiczna i w dodatku wartości są posortowane po kluczu. http://prac.im.pwr.wroc.pl/~szwabin/assets/algo/lectures/9.pdf

Reply:
Zależy co chcesz wyszukiwać. :) 

W przypadku mapy wyszukiwanie po kluczy jest również O(1) czyli czasie stałym. Oczywiście w najgorszym wypadku jest to O(n) ale ten najgorszy przypadek to sytuacja gdy źle zaimplementujemy hashcode i nie będzie równomiernego rozłożenia elementów.

Drzewa sa fajne, ale do innych operacji. Przykładowo - super odpowiadają na pytanie:
- Znajdź mi następnik/poprzednik elementu X

Ale "srednio" (bo logarytmicznie) odpowiadają na pytanie:
- Daj mi element dla klucza X